oximeter/db/tests/integration_test.rs

```diff
@@ -0,0 +1,380 @@
+
+    let client1 = Client::new_with_request_timeout(
+        deployment.http_addr(1)?,
+        log,
+        request_timeout,
+    );
+    let client2 = Client::new_with_request_timeout(
+        deployment.http_addr(2)?,
+        log,
+        request_timeout,
+    );
+    wait_for_ping(&client1).await?;
+    wait_for_ping(&client2).await?;
+    wait_for_keepers(&deployment, (1..=3).collect()).await?;
+    let end = tokio::time::Instant::now();
+    println!("deploy setup time = {:?}", end - start);
+
+    let start = tokio::time::Instant::now();
+    client1
+        .init_test_minimal_replicated_db()
+        .await
+        .context("failed to initialize db")?;
+    let end = tokio::time::Instant::now();
+    println!("init replicated db time = {:?}", end - start);
+
+    // Ensure our database tables show up on both servers
+    let start = tokio::time::Instant::now();
+    let output1 = client1.list_replicated_tables().await?;
+    let output2 = client2.list_replicated_tables().await?;
+    let end = tokio::time::Instant::now();
+    println!("list tables time = {:?}", end - start);
+    assert_eq!(output1, output2);
+
+    let input = TestInput::default();
+
+    // Let's write some samples to our first replica and wait for them to show
+    // up on replica 2.
+    let start = tokio::time::Instant::now();
+    let samples = test_util::generate_test_samples(
+        input.n_projects,
+        input.n_instances,
+        input.n_cpus,
+        input.n_samples,
+    );
+    let end = tokio::time::Instant::now();
+    println!("generate samples time = {:?}", end - start);
+    assert_eq!(samples.len(), input.n_points());
+    let start = tokio::time::Instant::now();
+    client1.insert_samples(&samples).await.expect("failed to insert samples");
+    let end = tokio::time::Instant::now();
+    println!("insert samples time = {:?}", end - start);
+
+    // Get all the samples from the replica where the data was inserted
+    let start = tokio::time::Instant::now();
+    let oxql_res1 = client1
+        .oxql_query("get virtual_machine:cpu_busy")
+        .await
+        .expect("failed to get all samples");
+    let end = tokio::time::Instant::now();
+    println!("query samples from client1 time = {:?}", end - start);
+
+    // Ensure the samples are correct on this replica
+    assert_input_and_output(&input, &samples, &oxql_res1);
+
+    let start = tokio::time::Instant::now();
+    wait_for_num_points(&client2, samples.len())
+        .await
+        .expect("failed to get samples from client2");
+    let end = tokio::time::Instant::now();
+    println!("query samples from client2 time = {:?}", end - start);
+
+    // Add a 3rd clickhouse server and wait for it to come up
+    deployment.add_server().expect("failed to launch a 3rd clickhouse server");
+    let client3 = Client::new_with_request_timeout(
+        deployment.http_addr(3)?,
+        log,
+        request_timeout,
+    );
+    wait_for_ping(&client3).await?;
```

Comment:
Based on reading `clickward`, this looks like it updates the config files for all the servers and keepers, right? So the servers should be re-reading those files and doing whatever they need to talk to the new node. Do we need to or want to wait here for not just the new replica to respond to pings, but for any acknowledgement that it's part of the cluster?

Reply:
I don't think so. That's what happens in the remainder of the test. We're essentially testing that the node becomes part of the cluster eventually and does what we expect. Without initializing the DB on the new node, the other nodes won't treat it as a replica yet anyway. We do that right below this line with `client3.init_test_minimal_replicated_db`